src/server/blocking_controller.cc

```diff
@@ -275,24 +276,29 @@ void BlockingController::AwakeWatched(DbIndex db_index, string_view db_key) {
 }
 
 // Marks the queue as active and notifies the first transaction in the queue.
-void BlockingController::NotifyWatchQueue(std::string_view key, WatchQueueMap* wqm) {
+void BlockingController::NotifyWatchQueue(std::string_view key, WatchQueueMap* wqm,
+                                          const DbContext& context) {
   auto w_it = wqm->find(key);
   CHECK(w_it != wqm->end());
   DVLOG(1) << "Notify WQ: [" << owner_->shard_id() << "] " << key;
   WatchQueue* wq = w_it->second.get();
 
-  DCHECK_EQ(wq->state, WatchQueue::SUSPENDED);
-  wq->state = WatchQueue::ACTIVE;
-
   auto& queue = wq->items;
   ShardId sid = owner_->shard_id();
 
-  do {
-    WatchItem& wi = queue.front();
-    Transaction* head = wi.get();
+  for (auto wi_it = queue.begin(), wi_e = queue.end(); wi_it != wi_e; ++wi_it) {
```

Comment:
as I said before, I would like to preserve a simple and efficient case of lists, sorted sets notifying a single waiter and stop.

Reply:
I don't change the behaviour)

Reply:
Didn't get your message at first. let's check how this implementation works. I can implement your suggestion regarding different queues for list, zset and stream like an optimization,